### PR TITLE
remove panic on unrecognized change

### PR DESCRIPTION
**Pull request by Mythra**
tl;dr

panics are super scary, probably don't need to panic just because
we barfed on one change. simply logging it, and returning should
be fine.

---

Motivation:

We were looking through the source of jsoncdc, and had noticed there was
a panic inside `append_change`. We couldn't think of a reason to panic
here (although we don't have as much insight I'm sure). Talking it over
we decided to switch to simply logging the unrecognized change action
instead.

We don't see a problem with replacing this panic (and all of the tests
still pass), however if you want we can change this PR to use a
compile time feature that isn't on by default.

Docker Changes:

I was unable to build the dockerfile when attempting to run tests.
Specifically it failed building `rustfmt`, by adding in gcc-multilib
I was able to build the dockerfile so I've added that into my change too.

### CODE REVIEW

**Review by posix4e — APPROVED** [review]
@solidsnack testify!

**Comment by posix4e** [comment]
Looks like the build failed

**Comment by Mythra** [comment]
Yea looks like one failed with:

```util/travis: line 41: LD_LIBRARY_PATH: unbound variable```
And the other failed with:
```Error: Refusing to uninstall /usr/local/Cellar/postgresql/9.5.4_1
because it is required by postgis, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies postgresql
```

Is this flaky installation, and a retrigger is needed? Should I look into changing the travis build script to get it to pass?

**Comment by posix4e** [comment]
Please feel free to rejigger

**Comment by Mythra** [comment]
Looks like they failed with the same errors again.

Although they don't look like things I've changed they due look like legit build failures (and travis has certainly updated their images).

I'll take a look into updating `utils/travis`

**Comment by Mythra** [comment]
Looks like this fails on Postgres 9.4 legitimately (after fixing travis), So I'm going to clean that up, and then test locally 9.4, 9.5, and 9.6 (instead of just 9.5)

File: src/lib.rs [hunk 1]
```diff
@@ -98,6 +99,26 @@ impl PGAppend<*mut i8> for pg::StringInfo {
...
+        let formatted_token = match self.0 {
+            REORDER_BUFFER_CHANGE_INSERT => "insert",
+            REORDER_BUFFER_CHANGE_UPDATE => "update",
+            REORDER_BUFFER_CHANGE_DELETE => "delete",
+            REORDER_BUFFER_CHANGE_INTERNAL_SNAPSHOT => "internal_snapshot",
+            REORDER_BUFFER_CHANGE_INTERNAL_COMMAND_ID => "internal_command_id",
+            REORDER_BUFFER_CHANGE_INTERNAL_TUPLECID => "internal_tuplecid",
+            REORDER_BUFFER_CHANGE_INTERNAL_SPEC_INSERT => "internal_spec_insert",
+            REORDER_BUFFER_CHANGE_INTERNAL_SPEC_CONFIRM => "internal_spec_confirm",
+            REORDER_BUFFER_CHANGE_MESSAGE => "change_message"
```
Comment (Mythra) [inline src/lib.rs]: So this is kind of weird, and I feel I should explain it.

This is a value in the postgres enum, but only for version 9.4. So on 9.4 this correctly generates all paths (and allows the code to be built).

On 9.5+ this acts as a "catch all" (this technically goes unused since we have all the previous values defined), but luckily rust still accepts it as a named `_` and allows the code to build.

Really once 9.4 support is no longer needed you can drop this.

Comment (solidsnack) [reply src/lib.rs]: So `REORDER_BUFFER_CHANGE_MESSAGE` is the catch all? Maybe it would be better to have a future proof `_`?

**Comment by solidsnack** [comment]
Please run `rustfmt` and repush. (Build failed on Linux already.)

**Comment by Mythra** [comment]
Heh doing that right now since I noticed it failed. I wanna make sure the mac build gets to that point though (since it was also failing). Then I'll repush.

File: src/lib.rs (hunk 1 above)
Comment (Mythra) [reply src/lib.rs]: So it is a catch all on 9.5, and 9.6 (not on 9.4).

On 9.5/9.6 Rust evaluates this as a `_` (since the value doesn't exist). However, it will also generate warnings that this code is unreachable, because we manually specify every single value. So on the places where it acts as a catch all it's impossible to hit.

However on 9.4 this generates all possible values for the enum. If this weren't here it would fail to build on 9.4 only since it didn't cover all the options.

If this were an underscore we would lose the ability to print out which change it was later on in the file (where we print unrecognized change action).

**Comment by Mythra** [comment]
I Think my local ~/.rustfmt.toml got in the way that time, but assuming this passes would you like me to squash these changes @solidsnack ?

**Comment by solidsnack** [comment]
If you were to squash and provide a comprehensive commit message that'd be great.

**Comment by Mythra** [comment]
Confused by what happened on this latest linux failure by travis.

Output shows:
```
make[1]: Leaving directory `/home/travis/build/posix4e/jsoncdc'
util/checkstyle
Not formatted as auto-formatting would: src/lib.rs
--- src/lib.rs.bk	2017-04-19 02:12:39.758051397 +0000
+++ src/lib.rs	2017-04-19 02:18:22.716023069 +0000
@@ -66,8 +66,8 @@
     pg::OutputPluginPrepareWrite(ctx, CTRUE);
     pg::appendStringInfo((*ctx).out, s.as_ptr(), (*txn).xid, t);
     pg::OutputPluginWrite(ctx, CTRUE);
-    let last_relid: *mut pg::Oid =
-        (*ctx).output_plugin_private as *mut pg::Oid;
+    let last_relid: *mut pg::Oid = (*ctx).output_plugin_private as
+                                   *mut pg::Oid;
```

Yet code shows:
```
    let last_relid: *mut pg::Oid = (*ctx).output_plugin_private as
                                    *mut pg::Oid;
```

Which should be exactly what it wants.

I'll squash down, and amend my original commit message to mention the travis changes. Although I'd be interested in why this is happening (since it doesn't happen locally in docker for me).

**Comment by solidsnack** [comment]
It's saying that auto-formatting would prefer this:

```
    let last_relid: *mut pg::Oid =
        (*ctx).output_plugin_private as *mut pg::Oid;
```

If you run `cargo fmt` it ensures the "project style" is followed.

**Comment by Mythra** [comment]
I'll change it to that, and try to figure out why a `cargo fmt` isn't applying that change (even still). Anyway I'll push that up and squash.

**Comment by Mythra** [comment]
Fixed up the formatting, and have squashed my commits.

Sorry about the number of commits as I tried to test travis, and the mistake about the rustfmt output.

**Review by solidsnack — APPROVED** [review]

**Comment by Mythra** [comment]
Now I'm really confused. Travis errors with:

```
make[1]: Leaving directory `/home/travis/build/posix4e/jsoncdc'
util/checkstyle
Not formatted as auto-formatting would: src/lib.rs
--- src/lib.rs.bk	2017-04-19 02:44:18.361701984 +0000
+++ src/lib.rs	2017-04-19 02:50:01.521701984 +0000
@@ -45,8 +45,8 @@
                             relation: pg::Relation,
                             change: *mut pg::ReorderBufferChange) {
     let relid = (*relation).rd_id;
-    let last_relid: *mut pg::Oid =
-        (*ctx).output_plugin_private as *mut pg::Oid;
+    let last_relid: *mut pg::Oid = (*ctx).output_plugin_private as
+                                   *mut pg::Oid;
make: *** [test] Error 1
```

Yet Looking at the code both locally (and on github: https://github.com/instructure/jsoncdc/blob/no-panic/src/lib.rs#L48 )

The code is what it wants:

```
    let last_relid: *mut pg::Oid =
        (*ctx).output_plugin_private as *mut pg::Oid;
```

**Comment by solidsnack** [comment]
Sometimes I think Travis gets stale stuff.

**Comment by Mythra** [comment]
:disappointed: I'll close/re-open to retrigger

**Comment by solidsnack** [comment]
Yeah, that Travis run is for 0fb6c68.

**Comment by solidsnack** [comment]
I think `rustfmt` had some regressions. I've filed a bug:

https://github.com/rust-lang-nursery/rustfmt/issues/1460

Had to reformat a bunch of stuff to get it to finally go through (2b6e346).

**Comment by solidsnack** [comment]
Thanks for your contribution. This should go into master in a couple of hours.

**Comment by Mythra** [comment]
No worries, thanks for sticking with me :+1: 

Let me know if I can do anything else.
